src/main/java/com/cloudbees/jenkins/support/impl/FileDescriptorLimit.java

```diff
@@ -0,0 +1,174 @@
+        if (channel == null) {
+            return "N/A: No connection to node.";
+        }
+        return channel.call(new GetUlimit());
+    }
+
+    /**
+     * * For slave machines.
+     */
+    private static final class GetUlimit implements Callable<String, RuntimeException> {
+        public String call() {
+            ByteArrayOutputStream bos = new ByteArrayOutputStream();
+            try {
+                getUlimit(bos);
+                getOpenFileDescriptorCount(bos);
+                listAllOpenFileDescriptors(bos);
+                return bos.toString("utf-8");
+            } catch (Exception e) {
+                return bos.toString();
+            }
+        }
+    }
+
+    /**
+     * * Using OperatingSystemMXBean, we can obtain the total number of open file descriptors.
+     */
+    private static void getOpenFileDescriptorCount(OutputStream os) throws UnsupportedEncodingException {
+        OperatingSystemMXBean operatingSystemMXBean = ManagementFactory.getOperatingSystemMXBean();
+        final PrintWriter writer = new PrintWriter(new OutputStreamWriter(os, "utf-8"), true);
+        if (operatingSystemMXBean instanceof UnixOperatingSystemMXBean) {
+            UnixOperatingSystemMXBean unixOperatingSystemMXBean = (UnixOperatingSystemMXBean) operatingSystemMXBean;
+            writer.println("Open File Descriptor Count: " + unixOperatingSystemMXBean.getOpenFileDescriptorCount());
+        }
+        writer.flush();
+    }
+
+    /**
+     * * List all the open file descriptors. For Unix systems this information can be obtained by
+     * * going to /proc/self/fd. This will translate self to the correct PID of the current java
+     * * process. Each file in the folder is a symlink to the location of the file descriptor.
+     */
+    private static void listAllOpenFileDescriptors(OutputStream os) throws IOException {
+        final PrintWriter writer = new PrintWriter(new OutputStreamWriter(os, "utf-8"), true);
+        String osName = System.getProperty("os.name").toLowerCase();
+
+        if (osName != null && !osName.contains("win")) { // If unix
+            writer.println();
+            writer.println("All open files");
+            writer.println("==============");
+            File[] files = new File("/proc/self/fd").listFiles();
+            for (File file : files) {
+                writer.println(file.getCanonicalPath());
+            }
+        }
+    }
+
+    /**
+     * * This method executes the command "bash -c ulimit -a" on the machine. If an exception is thrown log it
+     * * to file and continue.
+     */
+    private static void getUlimit(OutputStream os) throws UnsupportedEncodingException {
+        final PrintWriter writer = new PrintWriter(new OutputStreamWriter(os, "utf-8"), true);
+        String osName = System.getProperty("os.name").toLowerCase();
+
+        if (osName != null && !osName.contains("win")) {
+            ProcessBuilder builder = new ProcessBuilder("bash", "-c", "ulimit -a");
```

Comment:
@christ66 I wonder if we shouldn't be getting both the hard and soft limits as I have seen cases where the limits reported by `-a` are not the actual soft limits but rather the hard limits, so you fail to see e.g. the file descriptor limit being reached